TASK: Correctly mark nullable method parameters as nullable

Benjamin-K opened the pull request:
**Review instructions**

This PR will correctly mark implicit nullable types as nullable, which will remove deprecation warnings in PHP 8.4.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Benjamin-K committed:
TASK: Correctly mark nullable method parameters as nullable

See: https://wiki.php.net/rfc/deprecate-implicitly-nullable-types

dlubitz commented:
Thank you @Benjamin-K. This is great. But we need to check why the tests are failing.

Benjamin-K commented:
@dlubitz I fixed the PR already. The only thing it now complains about is that I updated the bracket style of functions:

```php
# before
function name() {}

# after
function name()
{
}
```

which is from the [PSR-12: Extended Coding Style](https://www.php-fig.org/psr/psr-12/)

dlubitz commented:
No, you changed it the other way around:

```diff
-    public function collectGarbage(): void
-    {
-    }
+    public function collectGarbage(): void {}
```

mhsdesign commented:
which i think is correct, we dont use this shortsyntax because its not worth it ^^ having empty method bodys is rare either way and psr 12 prohibits it:

> The opening brace MUST go on its own line, and the closing brace MUST go on the next line following the body.

To apply all style ci changes just use this :D

```
curl https://github.styleci.io/analyses/dj7o6p/diff | git apply -
```

Benjamin-K commented:
> No, you changed it the other way around:
> 
> ```diff
> -    public function collectGarbage(): void
> -    {
> -    }
> +    public function collectGarbage(): void {}
> ```

You're right, sry. Seems like vscode does not use the full PSR-12 set. I fixed this now.

dlubitz commented on Neos.Flow/Classes/Error/Debugger.php:
```diff
@@ -606,7 +607,7 @@ public static function getRecursionLimit(): int
  * @api
  */
-function var_dump($variable, string $title = null, bool $return = false, bool $plaintext = null)
+function var_dump($variable, string $title = null, bool $return = false, ?bool $plaintext = null)
```
```suggestion
function var_dump($variable, ?string $title = null, bool $return = false, ?bool $plaintext = null)
```

dlubitz commented on Neos.Flow/Classes/Configuration/ConfigurationSchemaValidator.php:
```diff
@@ -70,7 +71,7 @@ class ConfigurationSchemaValidator
      * @throws Exception\SchemaValidationException
      */
-    public function validate(string $configurationType = null, string $path = null, array &$loadedSchemaFiles = []): Result
+    public function validate(string $configurationType = null, ?string $path = null, array &$loadedSchemaFiles = []): Result
```
```suggestion
    public function validate(?string $configurationType = null, ?string $path = null, array &$loadedSchemaFiles = []): Result
```

dlubitz commented on Neos.Flow/Classes/Command/TypeConverterCommandController.php:
```diff
@@ -38,7 +39,7 @@ class TypeConverterCommandController extends CommandController
      * @return void
      */
-    public function listCommand(string $source = null, string $target = null)
+    public function listCommand(string $source = null, ?string $target = null)
```
```suggestion
    public function listCommand(?string $source = null, ?string $target = null)
```

dlubitz commented on Neos.Flow/Classes/Command/SignalCommandController.php:
```diff
@@ -37,7 +38,7 @@ class SignalCommandController extends CommandController
      * @return void
      */
-    public function listConnectedCommand(string $className = null, string $methodName = null): void
+    public function listConnectedCommand(string $className = null, ?string $methodName = null): void
```
```suggestion
    public function listConnectedCommand(?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/AnsiConsoleBackend.php:
```diff
@@ -83,7 +84,7 @@ public function open(): void
      * @return void
      */
-    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void
+    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void
```
```suggestion
    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/BackendInterface.php:
```diff
@@ -42,7 +43,7 @@ public function open(): void;
      * @api
      */
-    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void;
+    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void;
```
```suggestion
    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void;
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/ConsoleBackend.php:
```diff
@@ -79,7 +80,7 @@ public function open(): void
      * @api
      */
-    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void
+    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void
```
```suggestion
    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/FileBackend.php:
```diff
@@ -228,7 +229,7 @@ protected function rotateLogFile(): void
      * @api
      */
-    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void
+    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void
```
```suggestion
    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/JsonFileBackend.php:
```diff
@@ -29,7 +30,7 @@ class JsonFileBackend extends FileBackend
      * @return void
      */
-    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void
+    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void
```
```suggestion
    public function append(string $message, int $severity = LOG_INFO, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Classes/Backend/NullBackend.php:
```diff
@@ -43,7 +44,7 @@ public function open(): void
      * @api
      */
-    public function append(string $message, int $severity = 1, $additionalData = null, string $packageKey = null, string $className = null, string $methodName = null): void
+    public function append(string $message, int $severity = 1, $additionalData = null, string $packageKey = null, string $className = null, ?string $methodName = null): void
```
```suggestion
    public function append(string $message, int $severity = 1, $additionalData = null, ?string $packageKey = null, ?string $className = null, ?string $methodName = null): void
```

dlubitz commented on Neos.Flow.Log/Tests/Unit/Backend/AbstractBackendTest.php:
```diff
@@ -33,7 +34,7 @@ protected function setUp(): void
 			class ' . $this->backendClassName . ' extends \Neos\Flow\Log\Backend\AbstractBackend {
 				public function open(): void {}
-				public function append(string $message, int $severity = 1, $additionalData = NULL, string $packageKey = NULL, string $className = NULL, string $methodName = NULL): void {}
+				public function append(string $message, int $severity = 1, $additionalData = NULL, string $packageKey = NULL, string $className = NULL, ?string $methodName = NULL): void {}
```
```suggestion
				public function append(string $message, int $severity = 1, $additionalData = NULL, ?string $packageKey = NULL, ?string $className = NULL, ?string $methodName = NULL): void {}
```

dlubitz commented on Neos.Flow/Classes/Aop/Builder/AbstractMethodInterceptorBuilder.php:
```diff
@@ -96,7 +98,7 @@ protected function buildArraySetupCode(array $array): string
      * @return string The generated code to be used in an "array()" definition
      */
-    protected function buildMethodArgumentsArrayCode(string $className = null, string $methodName = null, bool $useArgumentsArray = false): string
+    protected function buildMethodArgumentsArrayCode(string $className = null, ?string $methodName = null, bool $useArgumentsArray = false): string
```
```suggestion
    protected function buildMethodArgumentsArrayCode(?string $className = null, ?string $methodName = null, bool $useArgumentsArray = false): string
```

dlubitz commented on Neos.Flow/Classes/Aop/Builder/AbstractMethodInterceptorBuilder.php:
```diff
@@ -156,7 +158,7 @@ protected function buildSavedConstructorParametersCode(string $className = null)
      * @return string PHP code to be used in the method interceptor
      */
-    protected function buildAdvicesCode(array $groupedAdvices, string $methodName = null, string $targetClassName = null, string $declaringClassName = null): string
+    protected function buildAdvicesCode(array $groupedAdvices, string $methodName = null, string $targetClassName = null, ?string $declaringClassName = null): string
```
```suggestion
    protected function buildAdvicesCode(array $groupedAdvices, ?string $methodName = null, string $targetClassName = null, ?string $declaringClassName = null): string
```

dlubitz commented on Neos.Flow/Classes/Aop/JoinPoint.php:
```diff
@@ -74,7 +75,7 @@ class JoinPoint implements JoinPointInterface
      * @param \Exception $exception The exception thrown (only used for After Throwing advices)
      */
-    public function __construct($proxy, string $className, string $methodName, array $methodArguments, Advice\AdviceChain $adviceChain = null, $result = null, \Exception $exception = null)
+    public function __construct($proxy, string $className, string $methodName, array $methodArguments, Advice\AdviceChain $adviceChain = null, $result = null, ?\Exception $exception = null)
```
```suggestion
    public function __construct($proxy, string $className, string $methodName, array $methodArguments, ?Advice\AdviceChain $adviceChain = null, $result = null, ?\Exception $exception = null)
```

dlubitz commented on Neos.Flow/Classes/Cli/ConsoleOutput.php:
```diff
@@ -140,7 +141,7 @@ public function outputFormatted(string $text = '', array $arguments = [], int $l
      * @param string $headerTitle
      */
-    public function outputTable(array $rows, array $headers = null, string $headerTitle = null): void
+    public function outputTable(array $rows, array $headers = null, ?string $headerTitle = null): void
```
```suggestion
    public function outputTable(array $rows, ?array $headers = null, ?string $headerTitle = null): void
```

dlubitz commented on Neos.Flow/Classes/Mvc/Routing/Dto/MatchResult.php:
```diff
@@ -39,7 +40,7 @@ final class MatchResult
      * @param RouteTags $tags
      */
-    public function __construct($matchedValue, RouteTags $tags = null, RouteLifetime $lifetime = null)
+    public function __construct($matchedValue, RouteTags $tags = null, ?RouteLifetime $lifetime = null)
```
```suggestion
    public function __construct($matchedValue, ?RouteTags $tags = null, ?RouteLifetime $lifetime = null)
```

dlubitz commented on Neos.Flow/Classes/Mvc/Routing/Dto/ResolveResult.php:
```diff
@@ -53,7 +54,7 @@ final class ResolveResult
      * @param RouteTags $tags
      */
-    public function __construct(string $resolvedValue, UriConstraints $uriConstraints = null, RouteTags $tags = null, RouteLifetime $lifetime = null)
+    public function __construct(string $resolvedValue, UriConstraints $uriConstraints = null, RouteTags $tags = null, ?RouteLifetime $lifetime = null)
```
```suggestion
    public function __construct(string $resolvedValue, ?UriConstraints $uriConstraints = null, ?RouteTags $tags = null, ?RouteLifetime $lifetime = null)
```

dlubitz commented on Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php:
```diff
@@ -114,7 +115,7 @@ public function getCachedMatchResults(RouteContext $routeContext)
      * @return void
      */
-    public function storeMatchResults(RouteContext $routeContext, array $matchResults, RouteTags $matchedTags = null, RouteLifetime $matchedLifetime = null)
+    public function storeMatchResults(RouteContext $routeContext, array $matchResults, RouteTags $matchedTags = null, ?RouteLifetime $matchedLifetime = null)
```
```suggestion
    public function storeMatchResults(RouteContext $routeContext, array $matchResults, ?RouteTags $matchedTags = null, ?RouteLifetime $matchedLifetime = null)
```

dlubitz commented on Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php:
```diff
@@ -152,7 +153,7 @@ public function getCachedResolvedUriConstraints(ResolveContext $resolveContext)
      * @return void
      */
-    public function storeResolvedUriConstraints(ResolveContext $resolveContext, UriConstraints $uriConstraints, RouteTags $resolvedTags = null, RouteLifetime $resolvedLifetime = null)
+    public function storeResolvedUriConstraints(ResolveContext $resolveContext, UriConstraints $uriConstraints, RouteTags $resolvedTags = null, ?RouteLifetime $resolvedLifetime = null)
```
```suggestion
    public function storeResolvedUriConstraints(ResolveContext $resolveContext, UriConstraints $uriConstraints, ?RouteTags $resolvedTags = null, ?RouteLifetime $resolvedLifetime = null)
```

dlubitz commented on Neos.Flow/Classes/Mvc/Routing/UriBuilder.php:
```diff
@@ -289,7 +290,7 @@ public function reset()
      * @throws \Neos\Flow\Http\Exception
      */
-    public function uriFor(string $actionName, array $controllerArguments = [], string $controllerName = null, string $packageKey = null, string $subPackageKey = null)
+    public function uriFor(string $actionName, array $controllerArguments = [], string $controllerName = null, string $packageKey = null, ?string $subPackageKey = null)
```
```suggestion
    public function uriFor(string $actionName, array $controllerArguments = [], ?string $controllerName = null, ?string $packageKey = null, ?string $subPackageKey = null)
```

dlubitz commented on Neos.Flow/Classes/Package/PackageManager.php:
```diff
@@ -881,7 +885,7 @@ protected function getPackageKeyFromManifest(array $manifest, $packagePath): str
      * @return string
      */
-    protected function derivePackageKey(string $composerName, string $packageType = null, string $packagePath = '', string $autoloadNamespace = null): string
+    protected function derivePackageKey(string $composerName, string $packageType = null, string $packagePath = '', ?string $autoloadNamespace = null): string
```
```suggestion
    protected function derivePackageKey(string $composerName, ?string $packageType = null, string $packagePath = '', ?string $autoloadNamespace = null): string
```

dlubitz commented on Neos.Flow/Classes/Persistence/Doctrine/Logging/SqlLogger.php:
```diff
@@ -35,7 +36,7 @@ class SqlLogger implements \Doctrine\DBAL\Logging\SQLLogger
      * @return void
      */
-    public function startQuery($sql, array $params = null, array $types = null)
+    public function startQuery($sql, array $params = null, ?array $types = null)
```
```suggestion
    public function startQuery($sql, ?array $params = null, ?array $types = null)
```

dlubitz commented on Neos.Flow/Classes/Persistence/Doctrine/Service.php:
```diff
@@ -196,7 +197,7 @@ public function getEntityStatus(): array
      * @return mixed
      */
-    public function runDql(string $dql, int $hydrationMode = \Doctrine\ORM\Query::HYDRATE_OBJECT, int $firstResult = null, int $maxResult = null)
+    public function runDql(string $dql, int $hydrationMode = \Doctrine\ORM\Query::HYDRATE_OBJECT, int $firstResult = null, ?int $maxResult = null)
```
```suggestion
    public function runDql(string $dql, int $hydrationMode = \Doctrine\ORM\Query::HYDRATE_OBJECT, ?int $firstResult = null, ?int $maxResult = null)
```

dlubitz commented on Neos.Flow/Classes/Validation/ValidatorResolver.php:
```diff
@@ -160,7 +161,7 @@ public function getBaseValidatorConjunction($targetClassName, array $validationG
      * @throws Exception\InvalidValidationOptionsException
      */
-    public function buildMethodArgumentsValidatorConjunctions($className, $methodName, array $methodParameters = null, array $methodValidateAnnotations = null)
+    public function buildMethodArgumentsValidatorConjunctions($className, $methodName, array $methodParameters = null, ?array $methodValidateAnnotations = null)
```
```suggestion
    public function buildMethodArgumentsValidatorConjunctions($className, $methodName, ?array $methodParameters = null, ?array $methodValidateAnnotations = null)
```

dlubitz commented on Neos.FluidAdaptor/Classes/Command/DocumentationCommandController.php:
```diff
@@ -45,7 +46,7 @@ class DocumentationCommandController extends CommandController
      * @return void
      */
-    public function generateXsdCommand(string $phpNamespace, string $xsdNamespace = null, string $targetFile = null, string $xsdDomain = ''): void
+    public function generateXsdCommand(string $phpNamespace, string $xsdNamespace = null, ?string $targetFile = null, string $xsdDomain = ''): void
```
```suggestion
    public function generateXsdCommand(string $phpNamespace, ?string $xsdNamespace = null, ?string $targetFile = null, string $xsdDomain = ''): void
```

dlubitz commented on Neos.FluidAdaptor/Tests/Unit/View/TemplatePathsTest.php:
```diff
@@ -452,7 +453,7 @@ public function expandGenericPathPatternDataProvider()
      * @param string $expectedResult
      */
-    public function expandGenericPathPatternTests($package, $subPackage, $controller, $format, $templateRootPath, array $templateRootPaths = null, $partialRootPath, array $partialRootPaths = null, $layoutRootPath, array $layoutRootPaths = null, $bubbleControllerAndSubpackage, $formatIsOptional, $pattern, $expectedResult)
+    public function expandGenericPathPatternTests($package, $subPackage, $controller, $format, $templateRootPath, array $templateRootPaths = null, $partialRootPath, array $partialRootPaths = null, $layoutRootPath, ?array $layoutRootPaths = null, $bubbleControllerAndSubpackage, $formatIsOptional, $pattern, $expectedResult)
```
```suggestion
    public function expandGenericPathPatternTests($package, $subPackage, $controller, $format, $templateRootPath, ?array $templateRootPaths = null, $partialRootPath, ?array $partialRootPaths = null, $layoutRootPath, ?array $layoutRootPaths = null, $bubbleControllerAndSubpackage, $formatIsOptional, $pattern, $expectedResult)
```

dlubitz commented on Neos.Http.Factories/Classes/UploadedFileFactoryTrait.php:
```diff
@@ -15,7 +16,7 @@ trait UploadedFileFactoryTrait
      * @inheritDoc
      */
-    public function createUploadedFile(StreamInterface $stream, int $size = null, int $error = \UPLOAD_ERR_OK, string $clientFilename = null, string $clientMediaType = null): UploadedFileInterface
+    public function createUploadedFile(StreamInterface $stream, int $size = null, int $error = \UPLOAD_ERR_OK, string $clientFilename = null, ?string $clientMediaType = null): UploadedFileInterface
```
```suggestion
    public function createUploadedFile(StreamInterface $stream, ?int $size = null, int $error = \UPLOAD_ERR_OK, ?string $clientFilename = null, ?string $clientMediaType = null): UploadedFileInterface
```

dlubitz reviewed (commented):
This is awesome. Thank you so much. 
Even though you found so many places I found some additional while reviewing.

dlubitz commented:
I also stumbled over this rector rule, which should do the migration: 
https://getrector.com/rule-detail/explicit-nullable-param-type-rector

dlubitz commented:
@Benjamin-K I can provide a PR with a rector migrated Flow which contains also the suggested changes. Is that OK for you?

Benjamin-K commented:
@dlubitz sure, feel free. If you did, I will close this PR.

dlubitz commented:
Done. Follow up PR is https://github.com/neos/flow-development-collection/pull/3429
